pypicloud/cache/dynamo.py

```diff
@@ -22,6 +25,78 @@
 LOG = logging.getLogger(__name__)
 
 
+def _path_depth(package):
+    """Count the slashes in path, or return infinity on no path."""
+    path = package.data.get('path')
+    if not path:
+        return float('+inf')
+    return path.count('/')
+
+
+def _decide_between_versions(contender, current):
+    """Decide between packages with the same filename on different paths."""
+
+    if contender.last_modified < current.last_modified:
+        return contender
```

Comment:
Not wasting time at all, we're figuring things out. The mere fact of having to put things down for another human has already convinced/dissuaded me of very important things!

I think I was unclear. What I meant is that you lead me to believe that was the right call, not what my patch actually does. Your assessment of what the code is currently doing is correct, and you are correct to suggest that we should compare pathness first and foremost.

To make sure we're on the same page, here's what I believe the code should look like:
```
def _decide_between_versions(cache, storage):
  if cache.filename != storage.filename:
    return current  # Trivial decision; pretty sure we agree on this.

  if cache.data['path'] == storage.data['path']:
    return contender  # Storage has new data for us on this path. Just accept it.

  # Here, we're dealing with two different files (since they have distinct paths)
  # with the same filename. Here are the two things I want to do:
  # * We should figure out a way to get them to decide on the same end state,
  #   regardless of who came first.
  # * We should always keep who comes first, in order to avoid surprises.
```

Ah! This is where I'm betraying myself. I want it both to be convergent and not-changing. This is incompatible (as I'm defining it right now). It's probably tolerable if it's a fraction of a millisecond (while two updates are trying to race each other), not so fine if it's during observable spans. That probably also explains why I confused you. I need some time to decide on a position to take here.

re what the cache is doing: not really (or it wasn't my intention). Cache rebuilds are supposed to make the cache reflect what we can observe in S3. Regardless of past state and events, given a snapshot of the bucket's contents, it should yield the same data on the cache. Sure, the mutation we do to it might be tailored to whatever is already in the cache, but the end state should always be the same (or converge to be the same, because eventual consistency and races).

This means that a "guided refresh" (by which I mean the informed approach in this patch) should yield the same results as an "absolute reset" (where we throw everything away and start with a fresh cache). That's what I was trying to advocate for.